### PR TITLE
Added custom privacy tagging ability

### PR DESCRIPTION
**Pull request by ssivov**

### CODE REVIEW

File: clarity.d.ts [hunk 1]
```diff
@@ -187,8 +187,9 @@ interface ICompressedBatchMessage extends IWorkerMessage {
 type NumberJson = Array<number | number[]>;
 
 interface IShadowDomNode extends HTMLDivElement {
-  node: Node; /* Reference to the node in the real DOM */
-  ignore: boolean;  /* Flag to avoid sending data for that node */
+  node: Node;
+  info: INodeInfo;
+  updateInfo: () => INodeInfo;
```
Comment (sarveshnagpal) [inline clarity.d.ts]: >updateInfo: () => INodeInfo; [](start = 2, length = 28)

What's the purpose of this function?

File: src/plugins/layout.ts [hunk 2]
```diff
@@ -88,9 +88,9 @@ export default class Layout implements IPlugin {
   }
 
   // Add node to the ShadowDom to store initial adjacent node info in a layout and obtain an index
-  private discoverNode(node: Node): ILayoutState {
-    this.shadowDom.insertShadowNode(node, getNodeIndex(node.parentNode), getNodeIndex(node.nextSibling));
-    return createLayoutState(node);
+  private discoverNode(node: Node): INodeInfo {
+    let shadowNode = this.shadowDom.insertShadowNode(node, getNodeIndex(node.parentNode), getNodeIndex(node.nextSibling));
+    return shadowNode.updateInfo();
```
Comment (sarveshnagpal) [inline src/plugins/layout.ts]: >.updateInfo() [](start = 21, length = 13)

Can't we used the cached info? I'm sure I am missing something here.. but also not intuitive why you would want to compute the info on the fly.

File: clarity.d.ts (hunk 1 above)
Comment (ssivov) [reply clarity.d.ts]: Gets an updated state of the node, including the new state

---
In reply to: [182825081](https://github.com/Microsoft/clarity-js/pull/80#discussion_r182825081) [](ancestors = 182825081)

File: src/plugins/layout.ts [hunk 3]
```diff
@@ -170,15 +171,11 @@ export default class Layout implements IPlugin {
 
   private layoutHandler(element: Element, source: Source) {
     let index = getNodeIndex(element);
+    let nodeInfo = this.shadowDom.getNodeInfo(index);
```
Comment (sarveshnagpal) [inline src/plugins/layout.ts]: >index [](start = 46, length = 5)

Can't index be null? We used to check earlier.. not anymore.

Comment (ssivov) [reply src/plugins/layout.ts]: If index is null, nodeinfo will be null too

---
In reply to: [182826132](https://github.com/Microsoft/clarity-js/pull/80#discussion_r182826132) [](ancestors = 182826132)

File: src/plugins/layout.ts (hunk 2 above)
Comment (ssivov) [reply src/plugins/layout.ts]: There is no cached info at this time. Info is not auto-generated on "insertShadowNode" call.

---
In reply to: [182825776](https://github.com/Microsoft/clarity-js/pull/80#discussion_r182825776) [](ancestors = 182825776)

File: src/plugins/layout/stateprovider.ts [hunk 4]
```diff
@@ -1,9 +1,10 @@
-import { IAttributes, IDoctypeLayoutState, IElementLayoutState, IIgnoreLayoutState, ILayoutState, ILayoutStyle, IStyleLayoutState,
-  ITextLayoutState } from "../../../clarity";
+import { IAttributes, IDoctypeLayoutState, IElementLayoutState, IIgnoreLayoutState, ILayoutRectangle, ILayoutState, ILayoutStyle,
+  INodeInfo, IStyleLayoutState, ITextLayoutState} from "../../../clarity";
```
Comment (sarveshnagpal) [inline src/plugins/layout/stateprovider.ts]: >INodeInfo [](start = 2, length = 9)

Redundant

File: src/plugins/layout.ts [hunk 5]
```diff
@@ -107,7 +107,8 @@ export default class Layout implements IPlugin {
 
   private createEventState<T extends ILayoutEventInfo>(eventInfo: T): ILayoutState {
     let node = eventInfo.node;
-    let layoutState: ILayoutState = createLayoutState(node);
+    let shadowNode = this.shadowDom.getShadowNode(eventInfo.index);
+    let layoutState: ILayoutState = shadowNode.updateInfo().state;
```
Comment (sarveshnagpal) [inline src/plugins/layout.ts]: >updateInfo [](start = 47, length = 10)

Nitpick: computeInfo

File: src/plugins/layout/stateprovider.ts [hunk 6]
```diff
@@ -100,8 +99,54 @@ export function createElementLayoutState(element: Element): IElementLayoutState
...
+  }
+  layoutState.cssRules = cssRulesTexts;
+  return layoutState;
+}
+
+export function createTextLayoutState(textNode: Text, maskText: boolean): ITextLayoutState {
+  // Text nodes that are children of the STYLE elements contain CSS code, so we don't want to hide it
+  // Checking parentNode, instead of parentElement, because in IE textNode.parentElement returns 'undefined'.
+  let parent = textNode.parentNode;
+  let isCss = parent && parent.nodeType === Node.ELEMENT_NODE && (parent as Element).tagName === "STYLE";
```
[the remark this answers is not in the document]
Comment (sarveshnagpal) [reply src/plugins/layout/stateprovider.ts]: Slava: won't fix.

---
In reply to: [182851714](https://github.com/Microsoft/clarity-js/pull/80#discussion_r182851714) [](ancestors = 182851714)

File: test/layout.ts [hunk 7]
```diff
@@ -1048,4 +1072,59 @@ describe("Layout Tests", () => {
...
+      let events = stopObserving();
+      assert.equal(events.length, 1);
+      assert.equal(events[0].state.tag, "INPUT");
+      assert.equal(events[0].state.action, Action.Insert);
+      assert.equal(events[0].state.attributes["value"], maskedValueString);
+      done();
+    }
+  });
+
+  it("checks that text value is masked when the config is set to show text and parent mask attribute is applied", (done: DoneFn) => {
```
Comment (sarveshnagpal) [inline test/layout.ts]: >that [](start = 13, length = 4)

Right now if a mutation adds "data-clarity-mask" child elements wouldn't automatically inherit the masking property. FYI. We can file it as an issue and come back.

Comment (ssivov) [reply test/layout.ts]: Added a test that currently fails and marked it as a bug

---
In reply to: [182854375](https://github.com/Microsoft/clarity-js/pull/80#discussion_r182854375) [](ancestors = 182854375)

**Pull request merged**